otter/log/spec.py

```diff
@@ -0,0 +1,84 @@
+"""
+Format logs based on specification
+"""
+
+from twisted.python.failure import Failure
+
+
+# mapping from msg type -> message
+msg_types = {
+    "execute-convergence": "Executing convergence",
+    "execute-convergence-results": (
+        "Got result of {worst_status} after executing convergence"),
+    "launch-servers": "Launching {num_servers} servers",
+    "delete-server": "Deleting {server_id} server",
+    "add-server-clb": ("Adding {server_id} with IP address {ip_address} "
+                       "to CLB {clb_id}"),
+    "remove-server-clb": ("Removing server {server_id} with IP address "
+                          "{ip_address} from CLB {clb_id}"),
+}
+
+
+def error_event(event, failure, why):
+    """
+    Convert event to error with failure and why
+    """
+    event["isError"] = True
+    event["failure"] = failure
+    event["why"] = why
+    event["original_message"] = event["message"]
+    event["message"] = ()
+    return event
+
+
+def SpecificationObserverWrapper(observer):
+    """
+    Return observer that validates messages based on specification
+    and delegates to given observer
+    """
+    def validating_observer(event_dict):
+        try:
+            speced_event = get_validated_event(event_dict)
+        except (ValueError, TypeError):
+            speced_event = error_event(
```

Comment:
It seems like this might lose more information than we ought to lose, by replacing the original `failure`, `why`, and `message` keys. Perhaps `error_event` could just produce an entirely new dict, with an `original_event` key which contains the entire original `event_dict`.